Trim whitespace of PDB residue names when building residue.id

sgalpha01 opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

Closes #3272

JoaoRodrigues commented on Bio/PDB/StructureBuilder.py:
```diff
@@ -118,7 +118,7 @@ def init_residue(self, resname, field, resseq, icode):
             if field == "H":
                 # The hetero field consists of H_ + the residue name (e.g. H_FUC)
-                field = "H_" + resname
+                field = ("H_" + resname).replace(" ", "")
```
I would rather use `str.strip` here than `replace`. More explicit. You can call it on the residue name only.

sgalpha01 replied:
Can I use `str.lstrip()`? It will be even more explicit. Or, should I simply use `str.strip()`?

JoaoRodrigues replied:
Do you know that spaces always come the left? :)

JoaoRodrigues commented:
Thanks! This change breaks some tests, so it'd de great if you could fix those too. It should involve only changing the expected results on some tests. The logs for appveyor and Travis should tell you what tests to look at.

I also edited your PR text to make use of the closing keywords (at the end), to link automatically to the issue.

JoaoRodrigues reviewed (commented):
Looks good! 

Going over the fix and the code I realized that also regular `ATOM` lines suffer from this problem. For instance, nucleotides (e.g. `DG`, `DC`) suffer from this same problem. In a structure coming from PDBParser, their residue names are ` DG` vs `DG` when parsed by MMCIFParser. We might tackle both issues by changing resname in `StructureBuilder` to `resname.strip()` the first time it comes through in `init_residue`. Then there's no need for this specific fix downstream.

Also, we should create new tests to check for this behavior. We already have two structures showing this 'problem' in our `Tests/PDB/` folder so we could simply write a new test in `test_PDB_SMCRA` that parses them both and compares them (chain by chain, residue by residue, atom by atom) to ensure they are the same. Would this be something you would be willing to do?

sgalpha01 commented:
@JoaoRodrigues , thanks for reviewing this. I will try my best to do the needed.

sgalpha01 commented:
> Looks good!
> 
> Going over the fix and the code I realized that also regular `ATOM` lines suffer from this problem. For instance, nucleotides (e.g. `DG`, `DC`) suffer from this same problem. In a structure coming from PDBParser, their residue names are ` DG` vs `DG` when parsed by MMCIFParser. We might tackle both issues by changing resname in `StructureBuilder` to `resname.strip()` the first time it comes through in `init_residue`. Then there's no need for this specific fix downstream.
> 
> Also, we should create new tests to check for this behavior. We already have two structures showing this 'problem' in our `Tests/PDB/` folder: `1LCD.pdb` and `1LCD.cif`. We could simply write a new test in `test_PDB_SMCRA` that parses them both and compares them (chain by chain, residue by residue, atom by atom) to ensure they are the same. Would this be something you would be willing to do?

Hello @JoaoRodrigues , I didn't understand the issue. Can you please elaborate more?
Also, did you mean the following changes:
``` Python3
...
        if field != " ":
            if field == "H":
                # The hetero field consists of H_ + the residue name (e.g. H_FUC)
                field = "H_" + resname.strip()
        res_id = (field, resseq, icode)
        if field == " ":
            if self.chain.has_id(res_id):
...
```
to something like:
``` Python3
...
        resname = resname.strip()  # Strips the whitespaces around resname

        if field != " ":
            if field == "H":
                # The hetero field consists of H_ + the residue name (e.g. H_FUC)
                field = "H_" + resname
        res_id = (field, resseq, icode)
        if field == " ":
            if self.chain.has_id(res_id):
...
```
I did so, but it broke two tests:
```
$ python run_tests.py --offline test_SeqIO_PdbIO.py
Skipping any tests requiring internet access
Python version: 3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
Operating system: nt win32
test_SeqIO_PdbIO ... FAIL
======================================================================
FAIL: test_atom_noheader (test_SeqIO_PdbIO.TestPdbAtom)
Parse a PDB with no HEADER line.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\Development\biopython\biopython\Tests\test_SeqIO_PdbIO.py", line 167, in test_atom_noheader
    self.assertEqual(len(chains), 1)
AssertionError: 3 != 1

----------------------------------------------------------------------
Ran 1 test in 5.078 seconds

FAILED (failures = 1)
```
and
```
$ python run_tests.py --offline test_SeqIO.py
Skipping any tests requiring internet access
Python version: 3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
Operating system: nt win32
test_SeqIO ... FAIL
======================================================================
FAIL: test_pdb_atom3 (test_SeqIO.TestSeqIO)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\Development\biopython\biopython\Tests\test_SeqIO.py", line 5463, in test_pdb_atom3
    self.perform_test(
  File "D:\Development\biopython\biopython\Tests\test_SeqIO.py", line 478, in perform_test
    self.assertEqual(
AssertionError: 3 != 1 : Found 3 records but expected 1

----------------------------------------------------------------------
Ran 1 test in 30.407 seconds

FAILED (failures = 1)
```

JoaoRodrigues commented:
Yes, I was even going to suggest placing that call to `.strip()` in `Bio/PDB/PDBParser.py` because it's the extra spaces are a 'feature' of the PDB format.

In summary, the PDB format is column-based so the residue name for each ATOM/HETATM record is located at columns 18, 19, and 20 (see [here for details](https://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#ATOM)). When you have a residue with a short name, such as `DG`, our parser still extracts those three characters from the line, making it ` DG` instead. Other formats (namely mmCIF) don't have this column-based format, so there are never empty spaces in the residue names. As we have it now, we cannot compare two structures easily because of issues like these. That's what I was trying to address with this issue. Is this clearer?

As for the test failures, this change might have consequences I did not antecipate. Both failures are in the `SeqIO` module. There's a type of reader (`pdb-atom`) that parses a PDB file and extracts the amino acid sequence. It's a bit of a rabbit hole, but the issue is with [this line](https://github.com/biopython/biopython/blob/master/Bio/SeqUtils/__init__.py#L307) in `Bio.SeqUtils.seq1`, which expects the input string to have a length that is divisible by 3. If you give it `DG` (length 2), it returns an empty list, which then eventually causes the error you are seeing. Since this is a very specific case, we can fix the call in `SeqIO/PdbIO.py` to workaround this issue. If you are not comfortable with fixing these other modules, just push the change to `PDBParser` and I can help you out in the other modules.

sgalpha01 commented:
Hey @JoaoRodrigues , now the issue is clearer to me. I've pushed the changes to `PDBParser.py`. And, thank you for being so cooperative.

Also, may I bother you again as I will definitely need help with the other modules?

JoaoRodrigues commented:
Hi @sgalpha01 of course, thanks for the fixes. Have a go at fixing the two failing tests and let us know if you need help. This isn't as easy as I thought it would be, sorry about that.

sgalpha01 commented:
I've created a test file, taking ideas from `test_SeqIO_PdbIO.py`, to check the issue.

test.py
``` Python3
import warnings

from Bio import BiopythonParserWarning, SeqIO
from Bio.PDB.PDBExceptions import PDBConstructionWarning


with warnings.catch_warnings():
    warnings.simplefilter("ignore", PDBConstructionWarning)
    warnings.simplefilter("ignore", BiopythonParserWarning)
    chains = list(SeqIO.parse("1LCD.pdb", "pdb-atom"))


print(chains)
```

This generates the following output:
``` Python3
[SeqRecord(seq=Seq('MKPVTLYDVAEYAGVSYQTVSRVVNQASHVSAKTREKVEAAMAELNYIPNR'), id='????:A', name='<unknown name>', description='????:A', dbxrefs=[]), 
SeqRecord(seq=Seq(''), id='????:B', name='<unknown name>', description='????:B', dbxrefs=[]), 
SeqRecord(seq=Seq(''), id='????:C', name='<unknown name>', description='????:C', dbxrefs=[])]
```

The test files can easily be resolved by modifying [this line](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Tests/test_SeqIO_PdbIO.py#L165) to `chains = list(SeqIO.parse("1LCD.pdb", "pdb-atom"))[:1]`. But, this doesn't solve the purpose. We need the chains list to be originally of length 1, right? I got the fact that this issue is coming due to the line you mentioned above, but can you give a little bit of clue about which module do I need to modify?

JoaoRodrigues commented:
Hi @sgalpha01 

I would suggest you follow the code/imports. Here's what I did when I checked:

1. That test calls `SeqIO.parse` with a file (`1LCD.pdb`) and a second argument `"pdb-atom"`. 
2. If you go to [Bio/SeqIO/__init__.py](https://github.com/biopython/biopython/blob/master/Bio/SeqIO/__init__.py) (since the import in the test is simply `from Bio import SeqIO`) and search for 'pdb-format' you see that it points to a class in a submodule: `Bio/SeqIO/PdbIO.PdbAtomIterator` [link](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L249). 
2. That function is actually quite short and besides parsing the PDB file, calls another function called `AtomIterator` defined at the top of the file, [here](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L20). 
3. [On line 54](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L54), there's an if statement that should skip processing a chain of the PDB file if there are no residues detected. So let's check how we find residues.
4. There's a list comprehension defined right before that if, [lines 49 to 53](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L49-L53) that uses a function called `seq1` to convert amino acid three-letter codes to one-letter codes: `seq1`. If the result of calling that function is _not_ "X", then the residue is accepted. Combining this with the if on line 54, we understand that `seq1` is the "gatekeeper" function here. So let's check it out.
5. `seq1` [is defined in Bio/SeqUtils/__init__.py](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqUtils/__init__.py#L256) and is quite short. Basically it expects a string or a list that has N elements, where N is a multiple of 3. Why? Because on line 307 you have a list comprehension that iterates over the input in strides of 3. Try running that function on something that is only 2 characters long, like `DG`. What do you get?
6. Now propagate that result to the other functions in `Bio/SeqIO`. Will that residues list be empty?

peterjc commented:
Somewhat related, see also #3125.

sgalpha01 commented:
Hi @JoaoRodrigues , I have pushed some changes. Not sure if it will break some other modules. Can you please check it?

JoaoRodrigues commented:
Hi @sgalpha01 

That's a good fix, but it's masking a bit the problem. 

What about we call `protein_letters_3to1` directly, instead of through `seq1`? Since in this loop we call `seq1` on one residue at a time, it's more efficient to call `protein_letters_3to1` (it's a dictionary) and use the `dict.get` function to return "X" if a match is not found. For consistency, we can make this change in the function `restype` that is defined on [line 39](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L39) and use `restype` in the loop on lines 49-53. 

We then need to check in the rest of `PdbIO.py` if there are other places where `seq1` gets called like this. If there are, it might be worth moving that `restype` function to the global scope (and maybe rename it to `_res2aacode` to be 1) private and 2) more explicit) so we can call it throughout the code.

What do you think?

sgalpha01 commented:
> Hi @sgalpha01
> 
> That's a good fix, but it's masking a bit the problem.
> 
> What about we call `protein_letters_3to1` directly, instead of through `seq1`? Since in this loop we call `seq1` on one residue at a time, it's more efficient to call `protein_letters_3to1` (it's a dictionary) and use the `dict.get` function to return "X" if a match is not found. For consistency, we can make this change in the function `restype` that is defined on [line 39](https://github.com/biopython/biopython/blob/36b5aee5652022d406e4a7529c1d23c525bf5d02/Bio/SeqIO/PdbIO.py#L39) and use `restype` in the loop on lines 49-53.
> 
> We then need to check in the rest of `PdbIO.py` if there are other places where `seq1` gets called like this. If there are, it might be worth moving that `restype` function to the global scope (and maybe rename it to `_res2aacode` to be 1) private and 2) more explicit) so we can call it throughout the code.
> 
> What do you think?

Yes, it would be a lot better. It will also increase the performance, ig. Okay, I'm on it.

sgalpha01 commented:
I've replaced `restype` function to `_res2aacode` with global scope. To make it almost similar to `seq1` function with single residue input, I've written a script to check difference between its default mapping and `protein_letters_3to1` as `custom_map`.

test.py
``` Python3

from Bio.Data.IUPACData import protein_letters_3to1_extended
from Bio.Data.SCOPData import protein_letters_3to1

onecode = {k.upper(): v for k, v in protein_letters_3to1_extended.items()}
for k, v in onecode.items():
    if k not in protein_letters_3to1:
        print(k, v)

```
It gives the following output:
```
XAA X
XLE J
PYL O
```
So, I've added these 3 mappings to the map in `_res2aacode` .

JoaoRodrigues commented on Bio/SeqIO/PdbIO.py:
```diff
@@ -17,6 +17,17 @@
+
+    Non-standard residues (e.g. CSD, ANP) are returned as 'X'.
+    """
+    onecode = protein_letters_3to1
```
You can simply make this `onecode = protein_letters_3to1_extended`, saves you the trouble of the update and retains exactly the same result as you'd have with `seq1`

sgalpha01 replied:
But, `protein_letters_3to1_extended` contains a lot lesser mappings than `protein_letters_3to1`. Will it be okay to do so?

sgalpha01 replied:
I've tried using `onecode = protein_letters_3to1_extended`, but it is causing tests to fail.

JoaoRodrigues replied:
That's interesting because it's what `seq1` uses internally. Can you double check that the errors come from this change and not from something else?

sgalpha01 replied:
Yes, I've checked. Initially, the `custom_map` argument was set to `protein_letters_3to1` in `restype` function while calling `seq1`. This combined both the dictionaries, and this combined dictionary was then used.

JoaoRodrigues replied:
Ah... I know what's up. This is confusing. We actually have two mappings called `protein_letters_3to1`, one in `Bio.Data.IUPACData` and one in `Bio.Data.SCOPData`. And one uses capitalized residue names while the other one does not.. This is messy. Let's keep it like this then, makes total sense.

JoaoRodrigues replied:
Actually, let's do this: we can import both and merge them at the top of the file:

```python
from Bio.Data.SCOPData import protein_letters_3to1 as scop_3to1
from Bio.Data.IUPACData import protein_letters_3to1_extended as iupac_3to1_ext

...

_aa3to1_dict = {}
_aa3to1_dict.update(iupac_3to1_ext)
_aa3to1_dict.update(scop_3to1)


def _res2aacode(...):  # uses _aa3to1_dict
    ...
```

JoaoRodrigues replied:
This way we get both dictionaries in, just like `seq1` and we keep them in sync with those dicts in other modules (no manual updates here).

sgalpha01 commented:
Okay, something is weird. I ran the tests on my local machine, and it passed them all.

JoaoRodrigues commented on Bio/SeqIO/PdbIO.py:
```diff
@@ -17,6 +17,17 @@
 
 
+def _res2aacode(residue, undef_code="X"):
+    """Return a residue's type as a one-letter code.
```
I would change the docstring to: "Return the one-letter amino acid code from the residue name.". Type here is not really appropriate.

JoaoRodrigues commented on Bio/SeqIO/PdbIO.py:
```diff
@@ -17,6 +17,17 @@
+def _res2aacode(residue, undef_code="X"):
+    """Return a residue's type as a one-letter code.
+
+    Non-standard residues (e.g. CSD, ANP) are returned as 'X'.
```
And I'd write: "Non-amino acid residues return X"

JoaoRodrigues reviewed (changes requested):
Very nice! I think we're almost there! Good job catching most of the calls to `seq1` and friends throughout the module!

JoaoRodrigues commented:
Let's wait for the tests - maybe we missed something.

JoaoRodrigues commented:
The difference is in the inputs for `seq1` and the `restype` functions. The first takes a string (essentially, `residue.resname`) and the second one takes a residue (and translates to `residue.resname` internally). There are two options: either we remember where to pass `.resname` and where not to, or we just add a small check to the `_res2aacode` function to handle the two inputs:

```python
def _res2aacode(residue, undef_code="X"):
    """Return a residue's type as a one-letter code.

    Non-standard residues (e.g. CSD, ANP) are returned as 'X'.
    """
    onecode = protein_letters_3to1
    onecode.update((k, v) for k, v in {"XAA": "X", "XLE": "J", "PYL": "O"}.items())

    if isinstance(residue, str):
        return onecode.get(residue, undef_code)

    return onecode.get(residue.resname, undef_code)
```

Since this is a private function, I think it's fine to fix this, what do you think?

sgalpha01 commented:
> The difference is in the inputs for `seq1` and the `restype` functions. The first takes a string (essentially, `residue.resname`) and the second one takes a residue (and translates to `residue.resname` internally). There are two options: either we remember where to pass `.resname` and where not to, or we just add a small check to the `_res2aacode` function to handle the two inputs:
> 
> ```python
> def _res2aacode(residue, undef_code="X"):
>     """Return a residue's type as a one-letter code.
> 
>     Non-standard residues (e.g. CSD, ANP) are returned as 'X'.
>     """
>     onecode = protein_letters_3to1
>     onecode.update((k, v) for k, v in {"XAA": "X", "XLE": "J", "PYL": "O"}.items())
> 
>     if isinstance(residue, str):
>         return onecode.get(residue, undef_code)
> 
>     return onecode.get(residue.resname, undef_code)
> ```
> 
> Since this is a private function, I think it's fine to fix this, what do you think?

I think the latter would be better. In this way, we don't have to remember where to pass `.resname`.

sgalpha01 commented:
Everything looks fine now

JoaoRodrigues commented on Bio/SeqIO/PdbIO.py:
```diff
@@ -11,12 +11,30 @@
+
+    Non-amino acid are returned as "X".
+    """
+    return (
```
I would prefer this being more explicit, but it's OK like this as well!

peterjc replied:
I agree, this version is five lines - an if/else would be four lines, and clearer.

JoaoRodrigues reviewed (approved):
I think this is it! Thank you so much for all the changes and edits, great job!

I'll defer to @peterjc on the changes to `SeqIO/PdbIO`. It seems OK to me, but I'm not super familiar with that module.

peterjc commented on Bio/SeqIO/PdbIO.py:
```diff
@@ -34,22 +52,13 @@ def AtomIterator(pdb_id, structure):
             res
             for res in chain.get_unpacked_list()
-            if seq1(res.get_resname().upper(), custom_map=protein_letters_3to1) != "X"
+            if _res2aacode(res.get_resname().upper(), undef_code="X") != "X"
```
Having the "X" explicit here and in the private function definition seems redundant and potentially confusing. Here you set the undefined code, but 20 lines later you do not.

I would suggest making the function take two required arguments, and then all the calls to it will be consistent. Or since you always use "X" just hard code it, no second argument?

peterjc reviewed (changes requested):
Some stylistic comments, but the changes in SeqIO look fine in principle.

JoaoRodrigues commented:
Thank you @sgalpha01 (and @peterjc  for reviewing)!
